sip/sip-002-smart-contract-language.md

```diff
@@ -98,48 +98,85 @@ any type.
 ## Inter-Contract Calls
 
 A smart contract may call functions from other smart contracts using a
-`(contract-call!)` function. This function accepts a function name and
-the smart contract's _identifier_ as input.  A smart contract's
-identifier is a hash of the smart contract's definition, represented
-as a Stacks address with a specific "smart contract" version byte. The
-smart contract identifier is a _principal_.
+`(contract-call?)` function.
 
-For example, to call the function `register-name` in a smart contract,
-you would use:
+This function returns a Response type result-- the return value of the called smart
+contract function. Note that if a called smart contract returns an
+`err` type, it is guaranteed to not alter any smart contract state
+whatsoever. Of course, any transaction fees paid for the execution
+of that function will not be returned.
+
+We distinguish 2 different types of `contract-call?`:
+
+* Static dispatch: the callee is a known, invariant contract available
+on-chain when the caller contract is being deployed. In this case, the
+callee's principal is provided as first argument, followed by the name
+of the method and its arguments:
 
 ```scheme
-(contract-call!
-    'SC3H92H297DX3YDPFHZGH90G8Z4NPH4VE8E83YWAQ
-    'register-name
+(contract-call?
+    'SC3H92H297DX3YDPFHZGH90G8Z4NPH4VE8E83YWAQ.registrar
+    register-name
     name-to-register)
 ```
 
-This function returns a boolean-- the return value of the called smart
-contract function. Note that if a called smart contract returns an
-`err` type, it is guaranteed to not alter any smart contract state
-whatsoever. Of course, any transaction fees paid for the execution
-of that function will not be returned.
+This approach must always be preferred, when adequat.
+It makes static analysis easier, and eliminates the
+potential for reentrancy bugs when the contracts are
+being published (versus when being used).
+
+* Dynamic dispatch: the callee is passed as an argument, and typed
+as a trait reference (<A>).
+
+```scheme
+(define-public (swap (token-a <can-transfer-tokens>)
+                     (amount-a uint)
+                     (owner-a principal)
+                     (token-b <can-transfer-tokens>)
+                     (amount-b uint)
+                     (owner-b principal)))
+     (begin
+         (unwrap! (contract-call? token-a transfer-from? owner-a owner-b amount-a))
+         (unwrap! (contract-call? token-b transfer-from? owner-b owner-a amount-b))))
+```
+
+Traits can either be locally defined:
+
+```scheme
+(define-trait can-transfer-tokens (
+    (transfer-from? (principal principal uint) (response uint)))
+```
+
+Or imported from an existing contract:
+
+```scheme
+(use-trait can-transfer-tokens
+    'SC3H92H297DX3YDPFHZGH90G8Z4NPH4VE8E83YWAQ.contract-defining-trait.can-transfer-tokens)
+```
+
+Looking at trait conformance, callee contracts have two different paths.
```

Comment:
I'm not sure about this.  Can we think of some real-world problems that cannot be solved if we require each contract to explicitly add `impl-trait` in order to be later invoked as a trait concretization?

Reply:
Yes -- I think that this is not so much about whether or not "accidental trait implementation" is absolutely required in all cases (it's not), but more about enabling a common case of implementation --- basically one smart contract defines a interface without knowing that it may be generally useful for other contracts, and then other contracts come along and implement it as a trait. If we forced explicit trait implementation, for the first contract to be used as a trait, it would need to relaunch, which isn't ideal.

It's true that implementing traits like this requires a dynamic check. However, it doesn't actually require a check at the site of the `contract-call?`, instead, it just need to be performed on `ClarityBlockConnection::run_contract_call`, since these concretizations can only be passed in as transaction arguments.

Reply:
Okay, you've convinced me that this is going to be a practical feature that solves real problems.

> It's true that implementing traits like this requires a dynamic check. However, it doesn't actually require a check at the site of the contract-call?, instead, it just need to be performed on ClarityBlockConnection::run_contract_call, since these concretizations can only be passed in as transaction arguments.

It looked like each `contract-call?` _within a contract_ was going to have to do a tedious and possibly very expensive runtime analysis check to determine whether or not the given contract was a valid concretization of the trait(s).  I just want to confirm that this isn't the case?

If we are certain that all concretization checks are performed at publish time for each `contract-call?` in a smart contract, and we're certain that all concretization type checks can be performed tractably for a `ContractCall` transaction, regardless of whether or not we use `(impl-trait)`, then this should be fine.

What's the purpose of `(impl-trait)`, then, if we're going to allow this "loose" trait concretization?  Does it make anything cheaper?  Asking because if we're happy with "loose" trait concretization, then `(impl-trait)` feels more like a hint for our _tooling_ to check that all trait methods are implemented.

Sorry if there are obvious answers to these questions -- this is the part of the codebase I'm least familiar with.

Reply:
Addressed with https://github.com/blockstack/blockstack-core/issues/1282